File: Answer/3741.三个相等元素之间的最小距离-ii.java
/*
 * @lc app=leetcode.cn id=3741 lang=java
 *
 * [3741] 三个相等元素之间的最小距离 II
 */

// @lc code=start
import java.util.HashMap;

class Solution {

    public int minimumDistance(int[] nums) {
        HashMap<Integer, State> positions = new HashMap<>();
        int best = Integer.MAX_VALUE;
        for (int index = 0; index < nums.length; index++) {
            int num = nums[index];
            State state = positions.get(num);
            if (state == null) {
                positions.put(num, new State(index));
                continue;
            }

            if (!state.hasTwo) {
                state.newer = index;
                state.hasTwo = true;
                continue;
            }

            best = Math.min(best, 2 * (index - state.older));
            state.older = state.newer;
            state.newer = index;
        }

        return best == Integer.MAX_VALUE ? -1 : best;
    }

    private static final class State {

        int older;
        int newer;
        boolean hasTwo;

        State(int index) {
            this.older = index;
        }
    }
}
// @lc code=end

File: Answer/3741.三个相等元素之间的最小距离-ii.py
#
# @lc app=leetcode.cn id=3741 lang=python3
#
# [3741] 三个相等元素之间的最小距离 II
#

# @lc code=start
from typing import List


class Solution:
    def minimumDistance(self, nums: List[int]) -> int:
        positions: dict[int, list[int]] = {}
        best = -1

        for index, num in enumerate(nums):
            state = positions.get(num)
            if state is None:
                positions[num] = [index]
                continue

            if len(state) == 1:
                state.append(index)
                continue

            if best == -1:
                best = 2 * (index - state[0])
            else:
                best = min(best, 2 * (index - state[0]))
            state[0] = state[1]
            state[1] = index

        return best
        
# @lc code=end

